whisper.cpp

```diff
@@ -1762,212 +1660,259 @@ static bool whisper_encode_internal(
+        if (!ggml_allocr_is_measure(alloc)) {
+            whisper_coreml_encode(wstate.ctx_coreml, (float *) mel->data, (float *) cur->data);
+        }
     }
 #endif
 #ifdef WHISPER_USE_OPENVINO
     else if (use_openvino) {
-        wstate.use_buf(ctx0, -1);
-
         cur = ggml_new_tensor_2d(ctx0, GGML_TYPE_F32, n_state, n_ctx);
+        ggml_allocr_alloc(alloc, cur);
 
-        if (!whisper_openvino_encode(wstate.ctx_openvino, mel, cur)) {
-            return false;
+        if (!ggml_allocr_is_measure(alloc)) {
+            whisper_openvino_encode(wstate.ctx_openvino, mel, cur);
         }
     }
 #endif
 
-    // cur
-    //{
-    //    printf("ne0 = %d\n", cur->ne[0]);
-    //    printf("ne1 = %d\n", cur->ne[1]);
-    //    for (int i = 0; i < 10; ++i) {
-    //        printf("%8.4f ", ((float *)(cur->data))[i]);
-    //    }
-    //    printf("... ");
-    //    for (int i = cur->ne[0] - 10; i < cur->ne[0]; ++i) {
-    //        printf("%8.4f ", ((float *)(cur->data))[i]);
-    //    }
-    //    printf("\n");
-    //}
+    ggml_build_forward_expand(gf, cur);
 
-    // pre-compute cross-attention memory
-    {
-        struct ggml_cgraph gf = {};
+    wstate.embd_enc = cur;
+
+    //ggml_graph_print(gf);
+
+    ////////////////////////////////////////////////////////////////////////////
+
+    //printf("%s: used_mem = %f MB, %f MB, %f MB %f MB %f MB\n", __func__,
+    //        ggml_used_mem(ctx0)/1024.0/1024.0,
+    //        wstate.get_buf_max_mem(0)/1024.0/1024.0,
+    //        wstate.get_buf_max_mem(1)/1024.0/1024.0,
+    //        wstate.get_buf_max_mem(2)/1024.0/1024.0,
+    //        wstate.get_buf_max_mem(3)/1024.0/1024.0);
+
+    ggml_free(ctx0);
+
+    return gf;
+}
+
+// pre-compute cross-attention memory
+static struct ggml_cgraph * whisper_build_graph_encoder_post(
+        whisper_context & wctx,
+          whisper_state & wstate) {
+    const auto & model   = wctx.model;
+    const auto & hparams = model.hparams;
+
+    const int n_ctx   = wstate.exp_n_audio_ctx > 0 ? wstate.exp_n_audio_ctx : hparams.n_audio_ctx;
+    const int n_state = hparams.n_audio_state;
+    const int n_head  = hparams.n_audio_head;
+
+    struct ggml_init_params params = {
+        /*.mem_size   =*/ wstate.buf_compute.size(),
+        /*.mem_buffer =*/ wstate.buf_compute.data(),
+        /*.no_alloc   =*/ true,
+    };
+
+    struct ggml_context * ctx0 = ggml_init(params);
 
-        // TODO: hack to disconnect the encoded features from the previous graph
-        cur->op = GGML_OP_NONE;
-        cur->src[0] = nullptr;
-        cur->src[1] = nullptr;
+    ggml_cgraph * gf = ggml_new_graph(ctx0);
 
-        for (int il = 0; il < model.hparams.n_text_layer; ++il) {
-            auto& layer = model.layers_decoder[il];
+    ggml_allocr * alloc = wstate.alloc_encode_post;
 
-            wstate.use_buf(ctx0, 0);
+    struct ggml_tensor * cur = ggml_view_tensor(ctx0, wstate.embd_enc);
```

Comment:
@slaren

I ported `ggml-alloc` into `whisper.cpp` and everything seems to work great.

There was one tricky moment, which I'm not sure why occurs - here it is in case you get some idea:

In Whisper, there are three steps executed one after the other:

- `encoder`
- `encoder_post`
- `decoder`

The `encoder` step outputs it's data into a tensor that we keep a reference to into the `whisper_state->embd_enc`:

https://github.com/ggerganov/whisper.cpp/blob/fa672b46e69e9aabd5b6799e0002372c4f8586eb/whisper.cpp#L1727-L1731

In the next step `encoder_post`, we want to use this data.
Initially, I directly tried to use the tensor from the state - i.e. `embd_enc`:

```c++
        struct ggml_tensor * cur = wstate.embd_enc;
        cur->op = GGML_OP_NONE;
        cur->src[0] = nullptr;
        cur->src[1] = nullptr;

       ...
```

And this seemed to work with the CPU version. Notice that this is not super clean, because `embd_enc` was created in the `ggml_ctx` of the `encoder` and this context is already destroyed. However, the allocator is alive and I checked that the `embd_enc->data` points to valid memory with the correct data.

Next I updated the CoreML functionality to work in the new implementation. The main difference when CoreML is enabled is that the `encoder` graph becomes just a single leaf in which the CoreML writes data:

https://github.com/ggerganov/whisper.cpp/blob/fa672b46e69e9aabd5b6799e0002372c4f8586eb/whisper.cpp#L1708-L1715

And this leaf is used in the same way in `encoder_post`.
However, this fails somewhere during the evaluation of `encoder_post` because the output data of `encoder_post` is incorrect. I checked that `embd_enc` has correct data at the start.

I thought that maybe because `embd_enc` is a tensor from a destroyed `ggml_ctx`, there is some problem that occurs. So tried to make a view of `embd_enc` which now lives is in the `ggml_ctx` of `encoder_post` and points to the same data:

https://github.com/ggerganov/whisper.cpp/blob/fa672b46e69e9aabd5b6799e0002372c4f8586eb/whisper.cpp#L1770-L1772

This is much cleaner compared to the hack above and it works correctly.
But still, I'm not sure why the first version does not work.

Reply:
I noticed that `whisper_build_graph_encoder` and `whisper_build_graph_encoder_post` seem to use the same buffer for the context, ie. `wstate.buf_compute`. So I think that what may be happening is that the tensor struct, which is allocated from this buffer, is overwritten in `whisper_build_graph_encoder_post`. When using `ggml_view_tensor`, a new tensor struct is allocated instead of trying to reuse the old one, so that fixes the issue. Could this be the cause?

Reply:
Yup, creating a separate `buf_compute` for `encoder_post` fixes the problem with the initial hacky version.